addons/path-mapped/common/src/main/java/org/commonjava/indy/pathmapped/cache/PathMappedMavenGACache.java

```diff
@@ -0,0 +1,387 @@
+        try
+        {
+            scan( notScanned, gaMap, completed );
+        }
+        catch ( Exception ex )
+        {
+            logger.error( "Failed to scan: ", ex );
+            return;
+        }
+
+        logger.debug( "Scan complete, completed: {}, gaMap: {}", completed, gaMap );
+        gaMap.forEach( ( ga, stores ) -> update( ga, stores ) );
+        update( SCANNED_STORES, completed );
+    }
+
+    private void update( String ga, Set<String> set )
+    {
+        BoundStatement bound = preparedStoresIncrement.bind();
+        bound.setSet( 0, set );
+        bound.setString( 1, ga );
+        session.execute( bound );
+        inMemoryCache.remove( ga ); // clear to force reloading
+    }
+
+    public void reduce( String ga, Set<String> set )
+    {
+        BoundStatement bound = preparedStoresReduction.bind();
+        bound.setSet( 0, set );
+        bound.setString( 1, ga );
+        session.execute( bound );
+        inMemoryCache.remove( ga ); // clear to force reloading
+    }
+
+    /**
+     * Scan the stores to find all GAs.
+     * @param notScanned
+     * @param gaMap
+     * @param completed
+     */
+    private void scan( final Set<String> notScanned, final Map<String, Set<String>> gaMap, final Set<String> completed )
+    {
+        PathDB pathDB = pathMappedFileManager.getPathDB();
+        notScanned.forEach( storeName -> {
+            Set<String> gaSet = new HashSet<>();
+            pathDB.traverse( "maven:hosted:" + storeName, ROOT_DIR, pathMap -> {
+                String gaPath = getGAPath( pathMap );
+                if ( isNotBlank( gaPath ) )
+                {
+                    gaSet.add( gaPath );
+                }
+            }, 0, PathDB.FileType.file );
+
+            gaSet.forEach( ga -> {
+                gaMap.computeIfAbsent( ga, k -> new HashSet<>() ).add( storeName );
+            } );
+            logger.info( "Scan result, store: {}, gaSet: {}", storeName, gaSet );
+            completed.add( storeName );
+        } );
+    }
+
+    /**
+     * Get GA path from pathMap obj. If it is pom file, get parent's parent.
+     */
+    private static String getGAPath( PathMap pathMap )
+    {
+        String ret = null;
+        String fileName = pathMap.getFilename();
+        if ( fileName.endsWith( ".pom" ) )
+        {
+            String parent = pathMap.getParentPath();
+            if ( isNotBlank( parent ) )
+            {
+                Path ga = Paths.get( parent ).getParent();
+                if ( ga != null )
+                {
+                    ret = ga.toString();
+                    if ( ret.startsWith( "/" ) )
+                    {
+                        ret = ret.substring( 1 ); // remove the leading '/'
+                    }
+                }
+            }
+        }
+        return ret;
+    }
+
+    public Set<String> getScannedStores()
+    {
+        return getStoresContaining( SCANNED_STORES );
```

Comment:
Is SCANNED_STORES the gaPath ?

Reply:
SCANNED_STORES is the special gaPath. It contains all scanned repos. We use it to check if a repo is "Unknown".

Reply:
Okay, see.